crates/ruff/src/rules/flake8_type_checking/rules/runtime_import_in_type_checking_block.rs

```diff
@@ -61,72 +67,168 @@ impl Violation for RuntimeImportInTypeCheckingBlock {
 /// TCH004
 pub(crate) fn runtime_import_in_type_checking_block(
```

Comment:
This, `typing_only_runtime_import`, and `unused_import` are structurally very similar. They all follow the same pattern:

- Iterate over individual imports, and group them by statement of origin.
- Iterate over the statements, and generate a single fix for all imports (taking into account suppression comments).
- Generate a diagnostic for each import in the statement, using that single fix.
- (Generate a diagnostic for each suppressed violation, so that the suppression comments aren't marked as unused.)

A lot of the complexity comes from handling suppression comments, since we need the unified fix to omit those members that are suppressed, and we rarely do that kind of thing in lint rules (we typically leave suppression to the next phase).

Also, while these three files are structurally similar, there are a bunch of nuanced differences between them that make me not want to bother DRYing them up much further. For example, in the typing-only import rule, we have to not only group by statement, but also by import _type_, since a single import statement can contain imports of multiple types (e.g., `import os, pandas`), which means we need a fix for each import _type_.